docs/core/diagnostics/distributed-tracing-instrumentation-walkthroughs.md

```diff
@@ -0,0 +1,412 @@
+is adding instrumentation for code in a 2nd independent assembly, the name should be based on the 
+assembly that defines the ActivitySource, not the assembly whose code is being instrumented.
+
+- The version parameter is optional. It is recommended to provide the version in case you release multiple
+versions of the library and make changes to the instrumented telemetry.
+
+> [!NOTE]
+> OpenTelemetry uses alternate terms 'Tracer' and 'Span'. In .NET 'ActivitySource' is the implementation
+> of Tracer and Activity is the implementation of 'Span'. .NET's Activity type long pre-dates
+> the OpenTelemetry specification and the original .NET naming has been preserved for
+> consistency within the .NET ecosystem and .NET application compatibility.
+
+### Activity Creation
+
+Use the ActivitySource object to Start and Stop Activity objects around meaningful units of work. Update
+DoSomeWork() with the code shown here:
+
+```csharp
+        static async Task DoSomeWork(string foo, int bar)
+        {
+            using (Activity activity = source.StartActivity("SomeWork"))
+            {
+                await StepOne();
+                await StepTwo();
+            }
+        }
+```
+
+Running the app now shows the new Activity being logged:
+```dotnetcli
+> dotnet run
+dotnet run
+Activity.Id:          00-9a41c8f2144cef45a99b9164db9da0b6-5cbc0f6836635245-01
+Activity.DisplayName: SomeWork
+Activity.Kind:        Internal
+Activity.StartTime:   2021-03-15T08:39:24.7379085Z
+Activity.Duration:    00:00:01.5198773
+Resource associated with Activity:
+    service.name: unknown_service:temp
+
+Example work done
+```
+
+#### Notes
+
+- <xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType> creates and starts
+the activity at the same time. The listed code pattern is using the `using` block which automatically disposes
+the created Activity object after executing the block. Disposing the Activity object will stop it so the code
+doesn't need to explicitly call <xref:System.Diagnostics.Activity.Stop?displayProperty=nameWithType>.
+That simplifies the coding pattern.
+
+- <xref:System.Diagnostics.ActivitySource.StartActivity?displayProperty=nameWithType> internally determines if
+there are any listeners recording the Activity. If there are no registered listeners or there are listeners which
+are not interested, `ActivitySource.StartActivity` will return `null` and avoid creating the Activity object. This
+is a performance optimization so that the code pattern can still be used in functions that are called very
+frequently.
+
+## Optional: Populate tags
+
+Activities support key-value data called Tags, commonly used to store any parameters of the work that
+may be useful for diagnostics. Update DoSomeWork() to include them:
+
+```csharp
+        static async Task DoSomeWork(string foo, int bar)
+        {
+            using (Activity activity = source.StartActivity("SomeWork"))
```

Comment:
Can we change this to Activity? so that it's more obvious this can return null?

```suggestion
            using (Activity? activity = source.StartActivity("SomeWork"))
```

Reply:
We can only use "Activity?" if we are using C# 8 which isn't the default version for a new console app in the 3.1 SDK. I assume we could force it forward but I worry it will generate compilation errors when people copy/paste the pattern into other projects. 

I'm planning to leave this as-is but leaving the issue open if others think we should switch it?